### ISSUE TITLE
Values file key deletion doesn't work

### ISSUE THREAD

**Issue by royjs** [helm/helm#3117]
I tried to use the feature to delete default keys (documented [here](https://docs.helm.sh/chart_template_guide/#deleting-a-default-key)) but it doesn't work.

To test it I created a simple values.yaml file like this;
```
a: a
b: b
c: c
```
and a small template:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data: |
{{- range $key, $value := .Values }}
  {{ $key }} - {{ $value }}
{{- end }}

```
When I try to delete the "b" key like this `helm install test-chart --debug --dry-run --set b=null` I get this:
```
COMPUTED VALUES:
a: 1
b: "null"
c: 3

HOOKS:
MANIFEST:

---
# Source: test-chart/templates/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data: |
  a - 1
  b - null
  c - 3
```

I also tried grabbing the source code and looking at the values and I think I found what part of the problem is. In the "coalesceValues" method there is a check for `value == nil` ([see here](https://github.com/kubernetes/helm/blob/648239edc3d2ee0adb607cecb859b013b7f3d86e/pkg/chartutil/values.go#L283)).  According to the [commit message](https://github.com/kubernetes/helm/pull/2648/commits/645f01eb22e03b5995b1b90834dc7321b4d813ed) it's expecting the value to be nil at this point but what I found out is that the value is actually a string with the value "null". So instead of deleting the key we actually end up using the key with a value of "null"

**Comment by jascott1** [comment]
I worked with @royjs on this and we determined it to be either a bug in implementation or explanation so labeling as 'bug'.

**Comment by RenaultAI** [comment]
I encountered the same bug - a parameter set to null gets evaluated as "null" instead of being unset.

The workaround I've found is by setting the parameter to false and using [flow control](https://github.com/kubernetes/helm/blob/master/docs/chart_template_guide/control_structures.md#ifelse) to unset the parameter.

config.yaml
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data: |
{{- range $key, $value := .Values }}
  {{- if $value }}
    {{ $key }} - {{ $value }}
  {{- end }}
{{- end }}
```
helm install output:
```
helm install --debug --dry-run --set b=false .
[debug] Created tunnel using local port: '35677'

[debug] SERVER: "127.0.0.1:35677"

[debug] Original chart version: ""
[debug] CHART PATH: /tmp/test-chart

NAME:   agile-turtle
REVISION: 1
RELEASED: Tue Jan 23 18:42:52 2018
CHART: test-chart-0.1.0
USER-SUPPLIED VALUES:
b: false

COMPUTED VALUES:
a: a
b: false
c: c

HOOKS:
MANIFEST:

---
# Source: test-chart/templates/config.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
data: |
    a - a
    c - c
```

**Comment by bonifaido** [comment]
We have encountered the same issue here: https://github.com/banzaicloud/banzai-charts/issues/214

**Comment by dannyk81** [comment]
I've just hit this in both v2.9.1 and v2.10.0, this is still not working as described.

```
$ helm template --set controller.nodeSelector."node-role\.kubernetes\.io\/ingress"=null
```

Result:
```
      nodeSelector:
        node-role.kubernetes.io/ingress: null
```


I tied setting an empty string instead, similar issue:

```
$ helm template --set controller.nodeSelector."node-role\.kubernetes\.io\/ingress"=""
```

Result:
```
      nodeSelector:
        node-role.kubernetes.io/ingress: ""
```

**Comment by Josh-West-Bose** [comment]
I've hit this bug as well.

**Comment by stealthybox** [comment]
I can confirm this issue.
My use case is to remove annotations in previous overrides.

```bash
# this sets the key to the string "null"
helm install test-chart --debug --dry-run --set b=null
```

YAML values files behave differently, as the commandline parsing is not involved, so you can actually use null:

```bash
# these null the values for the keys which is good, but the keys are not deleted as documented

echo "b: " | helm install test-chart --debug --dry-run --values /dev/stdin
echo "b: null" | helm install test-chart --debug --dry-run --values /dev/stdin
echo "b: ~" | helm install test-chart --debug --dry-run --values /dev/stdin
```
http://yaml.org/type/null.html

If you just need to null the value, this might be your workaround.
Unfortunately, this doesn't work for my use case as null can't be used for annotations:
```
Invalid value: "failed validation": unknown object type "nil" in Service.metadata.annotations.test
```

Also, when using `if` in go templates, both `'""` and `nil` fail, and there's no function to distinguish between them.
This makes it impractical to template arbitrary annotations from `.Values` without disallowing the empty string or some other special string such as `"-"`.

We should fix this so that it drops the key.

**Comment by stealthybox** [comment]
@marespane I'm working on a series of patches that fix several bugs with the value merging behavior in pkg/chartutil

**Comment by vsliouniaev** [comment]
There is some additional unexpected stateful behaviour:
If the resource looks like this
```
selector:
    a: ""
    b: 2
```
Then no matter what I do I cannot remove the `a` key

If the resource looks like this:
```
selector:
  b: 2
```
Then the same configuration will not add the `a=""` key.

This looks consistent with others' descriptions above, and is an issue for Helm's declarative approach. 
Since I hit this working with a `service` this caused the selector to not work properly depending on the heritage of the resource.

**Comment by swistaczek** [comment]
I have the same issue described by @sfitts

**Comment by kellyfj** [comment]
I have the same issue in

`$ helm version
Client: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}`

**Comment by stealthybox** [comment]
I still have a working patch for this -- I haven't found the time to push it over the last mile.
I found some other bugs along the way that after discussing with @mattfarina I should revert the fixes for.

**Comment by bacongobbler** [comment]
The original issue

> So instead of deleting the key we actually end up using the key with a value of "null"

Was fixed in #3912 which was released as part of helm 2.10.0, so @kellyfj you'll need to upgrade to get the fix.

If anyone else is running into an issue other than what @royjs reported in the OP, please open a new ticket. Thanks!

**Comment by kellyfj** [comment]
Many Thanks!
